// mocha/lib/cli/config.js
'use strict';

/**
 * Responsible for loading / finding Mocha's "rc" files.
 *
 * @private
 * @module
 */

var fs = require('fs');
var path = require('path');
var debug = require('debug')('mocha:cli:config');
var findUp = require('find-up');

var _require = require('../errors'),
    createUnparsableFileError = _require.createUnparsableFileError;

var utils = require('../utils');

/**
 * These are the valid config files, in order of precedence;
 * e.g., if `.mocharc.js` is present, then `.mocharc.yaml` and the rest
 * will be ignored.
 * The user should still be able to explicitly specify a file.
 * @private
 */
exports.CONFIG_FILES = ['.mocharc.cjs', '.mocharc.js', '.mocharc.yaml', '.mocharc.yml', '.mocharc.jsonc', '.mocharc.json'];

/**
 * Parsers for various config filetypes. Each accepts a filepath and
 * returns an object (but could throw)
 */
var parsers = exports.parsers = {
  yaml: function yaml(filepath) {
    return require('js-yaml').load(fs.readFileSync(filepath, 'utf8'));
  },
  js: function js(filepath) {
    var cwdFilepath = void 0;
    try {
      debug('parsers: load cwd-relative path: "%s"', path.resolve(filepath));
      cwdFilepath = require.resolve(path.resolve(filepath)); // evtl. throws
      return require(cwdFilepath);
    } catch (err) {
      if (cwdFilepath) throw err;

      debug('parsers: retry load as module-relative path: "%s"', filepath);
      return require(filepath);
    }
  },
  json: function json(filepath) {
    return JSON.parse(require('strip-json-comments')(fs.readFileSync(filepath, 'utf8')));
  }
};

/**
 * Loads and parses, based on file extension, a config file.
 * "JSON" files may have comments.
 *
 * @private
 * @param {string} filepath - Config file path to load
 * @returns {Object} Parsed config object
 */
exports.loadConfig = function (filepath) {
  var config = {};
  debug('loadConfig: trying to parse config at %s', filepath);

  var ext = path.extname(filepath);
  try {
    if (ext === '.yml' || ext === '.yaml') {
      config = parsers.yaml(filepath);
    } else if (ext === '.js' || ext === '.cjs') {
      config = parsers.js(filepath);
    } else {
      config = parsers.json(filepath);
    }
  } catch (err) {
    throw createUnparsableFileError('Unable to read/parse ' + filepath + ': ' + err, filepath);
  }
  return config;
};

/**
 * Find ("find up") config file starting at `cwd`
 *
 * @param {string} [cwd] - Current working directory
 * @returns {string|null} Filepath to config, if found
 */
exports.findConfig = function () {
  var cwd = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : utils.cwd();

  var filepath = findUp.sync(exports.CONFIG_FILES, { cwd: cwd });
  if (filepath) {
    debug('findConfig: found config file %s', filepath);
  }
  return filepath;
};
